ethcore/src/private_transactions/encryptor.rs

```diff
@@ -0,0 +1,300 @@
+			base_url: None,
+			threshold: 0,
+			sessions: Mutex::new(HashMap::new()),
+		})
+	}
+
+	/// Create new encryptor
+	pub fn new(key_server_url: Option<String>, threshold: u32) -> Result<Self, EthcoreError> {
+		Ok(SecretStoreEncryptor {
+			client: FetchClient::new()
+				.map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(format!("{}", e))))?,
+			base_url: key_server_url,
+			threshold: threshold,
+			sessions: Mutex::new(HashMap::new()),
+		})
+	}
+
+	/// Ask secret store for key && decrypt the key.
+	fn retrieve_key(
+		&self,
+		url_suffix: &str,
+		method: FetchMethod,
+		contract_address: &Address,
+		contract_address_signature: &Signature,
+		requester: &Address,
+		unlocked_accounts: Arc<AccountProvider>
+	) -> Result<Bytes, EthcoreError> {
+		// check if the key was already cached
+		if let Some(key) = self.obtained_key(contract_address) {
+			return Ok(key);
+		}
+		// key id in SS is H256 && we have H160 here => expand with assitional zeros
+		let mut contract_address_extended = Vec::with_capacity(32);
+		contract_address_extended.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+		contract_address_extended.extend_from_slice(&**contract_address);
+
+		let base_url = self.base_url.clone().ok_or_else(|| PrivateTransactionError::KeyServerNotSet)?;
+
+		// prepare request url
+		let url = format!("{}/{}/{}{}",
+				base_url,
+				contract_address_extended.to_hex(),
+				contract_address_signature,
+				url_suffix,
+			);
+
+		// send HTTP request
+		let mut response = self.client.fetch_with_abort(&url, method, Default::default()).wait()
+			.map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(format!("{}", e))))?;
+
+		if response.is_not_found() {
+			return Err(EthcoreError::PrivateTransaction(PrivateTransactionError::EncryptionKeyNotFound(*contract_address)));
+		}
+
+		if !response.is_success() {
+			return Err(EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(response.status().canonical_reason().unwrap_or("unknown").into())));
+		}
+
+		// read HTTP response
+		let mut result = String::new();
+		response.read_to_string(&mut result)?;
+
+		// response is JSON string (which is, in turn, hex-encoded, encrypted Public)
+		let result_len = result.len();
+		if result_len == 0 || &result[0..1] != "\"" || &result[result_len - 1..result_len] != "\"" {
+			return Err(EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(format!("Invalid SecretStore response: {}", result))));
+		}
+		let encrypted_bytes: ethjson::bytes::Bytes = result[1..result_len-1].parse().map_err(|e| EthcoreError::PrivateTransaction(PrivateTransactionError::Encrypt(e)))?;
+
+		// decrypt Public
+		let decrypted_bytes = unlocked_accounts.decrypt(*requester, None, &ethcrypto::DEFAULT_MAC, &encrypted_bytes)?;
+		let decrypted_key = Public::from_slice(&decrypted_bytes);
+
+		// and now take x coordinate of Public as a key
+		let key: Bytes = (*decrypted_key)[..INIT_VEC_LEN].into();
```

Comment:
What if it's shorter than `INIT_VEC_LEN`?

Reply:
`decrypted_key` is `Public` (64-bytes) and `INIT_VEC_LEN` is only 16 (bytes). Don't think additional check is required, because both code (range access) and constant are module-local.